executor/src/lib.rs

```diff
@@ -50,6 +50,41 @@ pub trait ExecutionState {
     /// The error type to return in case something went wrong during execution.
     type Error: ExecutionStateError;
 
+    /// Simple guardrail ensuring there is a single instance using the state
+    /// to call `handle_consensus_transaction`. Many instances may read the state,
+    /// or use it for other purposes.
+    fn ask_consensus_write_lock(&self) -> bool;
+
+    /// Tell the state that the caller instance is no longer using calling
+    //// `handle_consensus_transaction`.
+    fn release_consensus_write_lock(&self);
+}
+
+/// Execution state that gets whole certificates and the corresponding batches
+/// for execution. It is responsible for deduplication in case the same certificate
+/// is re-delivered after a crash.
+#[async_trait]
+pub trait BatchExecutionState: ExecutionState {
+    /// Load the last consensus index from storage.
+    ///
+    /// It should return the index from which it expects a replay, so one higher than
+    /// the last certificate index it successfully committed. This is so it has the
+    /// same semantics as `ExecutionIndices`.
+    async fn load_next_certificate_index(&self) -> Result<SequenceNumber, Self::Error>;
+
+    /// Execute the transactions and atomically persist the consensus index.
+    ///
+    /// TODO: This function should be allowed to return a new committee to reconfigure the system.
```

Comment:
Reconfiguration can simply be done by sending the following message to the Primary and the Workers through the network: `ReconfigureNotification::NewCommittee(new_committee)`.

Below is an example:
https://github.com/MystenLabs/narwhal/blob/ad05db4c5d6f851402efc7d2b285e9b59d513e64/node/src/restarter.rs#L98-L101

Reply:
Thanks for the link! True, we discussed that we can send reconfiguration messages through gRPC. 

I included this TODO because the [current docs](https://github.com/MystenLabs/narwhal/blob/52dd0c4efa36ae303b5ef80e015c108a2fb6955d/executor/src/lib.rs#L57-L58) of the `handle_consensus_transaction` say that it can return a new committee. While that's not true, as it clearly cannot, I thought the idea that it should do so was a good one. It capture committee transitions on the level of the `ExecutionState`, if it is to happen as part of transaction execution. The gRPC way seems like a technical detail that the implementors of the `ExecutionState` should not have to concern themselves with.

But it's up to you, I just wanted it to be consistent with the other docs.

Reply:
Cool, let's leave the TODO then (we haven't fully deployed Narwhal reconfiguration in DevNet yet, so this part may still change)